#!/usr/bin/env cwl-runner

cwlVersion: v1.0

class: CommandLineTool

requirements:
  - class: InlineJavascriptRequirement

  - class: EnvVarRequirement
    envDef:
      -
        envName: SENTIEON_LICENSE
        envValue: LICENSEID

hints:
  - class: DockerRequirement
    dockerPull: ACCOUNT/tnhaplotyper2:VERSION

baseCommand: [sentieon_merge_TNfilter_normal.sh]

inputs:
  - id: genome_reference_fasta
    type: File
    secondaryFiles:
      - ^.dict
      - .fai
    inputBinding:
      position: 1
    doc: Genome reference in FASTA format with the corresponding index files

  - id: sample_name
    type: string
    inputBinding:
      position: 2
    doc: Name of the sample

  - id: normal_name
    type: string
    inputBinding:
      position: 3
    doc: Name of the normal

  - id: input_files_vcf_gz
    type:
      -
        items: File
        type: array
    inputBinding:
      position: 4
    secondaryFiles:
      - .tbi
      - .stats
    doc: List of input VCF shards with the corresponding index and stats files

  - id: input_files_priors
    type:
      -
        items: File
        type: array
    inputBinding:
      position: 5
    doc: List of input OrientationBias shards

  - id: input_files_contamination
    type:
      -
        items: File
        type: array
    inputBinding:
      position: 6
    doc: List of input ContaminationModel shards

outputs:
  - id: output_file_vcf_gz
    type: File
    outputBinding:
      glob: output.vcf.gz
    secondaryFiles:
      - .tbi
      - .stats

doc: |
  Merge TNhaplotyper2, OrientationBias, and ContaminationModel |
  files generated by shards on tumor-normal data. |
  Run TNfilter to generate filtered calls in VCF format
